src/hp/magicsock/conn.rs

```diff
@@ -1138,23 +1139,37 @@ impl Actor {
             }
         };
 
-        let meta = quinn_udp::RecvMeta {
-            len: dm.buf.len(),
-            stride: dm.buf.len(),
-            addr: ep_fake_wg_addr,
-            // Normalize local_ip
-            dst_ip: self.normalized_local_addr().ok().map(|addr| addr.ip()),
-            ecn: None,
-        };
+        // the derp packet is made up of multiple udp packets, prefixed by a u16 be length prefix
+        //
+        // split the packet into these parts
+        let parts = PacketSplitIter::new(dm.buf);
+        // Normalize local_ip
+        let dst_ip = self.normalized_local_addr().ok().map(|addr| addr.ip());
+        let addr = ep_fake_wg_addr;
+        let res = parts
+            .map(|part| match part {
+                Ok(part) => {
+                    let meta = quinn_udp::RecvMeta {
+                        len: part.len(),
+                        stride: part.len(),
+                        addr,
+                        dst_ip,
+                        ecn: None,
+                    };
+                    NetworkReadResult::Ok {
+                        source: NetworkSource::Derp,
+                        bytes: part,
+                        meta,
+                    }
+                }
```

Comment:
hi! If we are splitting the packets up in `magicsock::Conn`, then I think we should be combining them in `magicsock::Conn` as well. 

We are currently (technically) changing the derp protocol by pushing the batching into `derp` itself. If we instead do that batching inside `magicsock::Conn` in `send_derp`, & pass `derp_client.send` the batched bytes, then we aren't breaking the protocol & are also treating the packets the same on the way in and on the way out.

If we do decide that we would rather break from the protocol & should be batching the packets inside derp itself, that I think we should do the "unbatching" inside derp as well, and pass out a vec of bytes, instead of a single blob that `magicsock` is responsible for unbatching.

But to be clear, I'm advocating we do the batching and unbatching inside `magicsock::send_derp` & `magicsock::process_derp_read_result` instead of changing `derp` code at all.

Reply:
I am happy to move stuff around. I just did this in the way that required the minimum amount of changes.

When you say breaking the derp protocol, you mean that we break the internal layering between the socket part and the derp part, right? Because in terms of the bytes we send over the wire I don't think this breaks anything. We are still sending `<peerid><bytes>`, just that <bytes> now is the concatenation of multiple udp transmits.